<a href="https://colab.research.google.com/github/HannahParker/cse480-notebooks/blob/master/11_2_Ponder_and_Prove_NP_Completeness.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove NP-Completeness
## Due: Saturday, 20 March 2021, 11:59 pm

## TODO Watch and Learn

Watch each of the following videos:

* [P vs. NP - The Biggest Unsolved Problem in Computer Science](https://www.youtube.com/watch?v=EHp4FPyajKQ)
* [What is complexity theory? (P vs. NP explained visually)](https://www.youtube.com/watch?v=u2DLlNQiPB4)
* [P vs. NP and the Computational Complexity Zoo](https://www.youtube.com/watch?v=YX40hbAHx3s&t=3s)

For each video, state two things that interested you (theoretical or practical). Just a couple of sentences that summarize each point of interest.

**P vs. NP - The Biggest Unsolved Problem in Computer Science**

- This was the first video I watched and thus, I learned what P vs. NP is. Problems are grouped in classes based on complelexity. P and NP are two of these classes. P and NP both take a very long time to calculate. NP: Easy to Check, Hard to Solve. P: Easy to Solve.

- Complexity classes are not entierly seperate form each other. P is actually inside of the NP class. However, NP complete problems has never been shown to conain P problems. Thus, P = NP really means P = NP complete. 

**What is complexity theory? (P vs. NP explained visually**

- Compexity theory includes looking at the time it takes to solve the problem and how the input size grows in relation to machine operations over time (or state transitions). This allows us to create a graph that shows the input size vs. time(steps).

- Most problems that we ask computers to solve are in Polynomial time. Other probelms use Exponential time. Short cuts have been found to compute exponential time solutions to polynomial time problems.

**P vs. NP and the Computational Complexity Zoo**

- Some problems can go from NP to P. Meaning a short cut can be found to find a faster solution to the problem. P vs. NP refers to the question, can all NP problems be converted to P problems?

- Examples of NP: Soduko, Crossword Puzzles
- Examples of P: Multiplication, Mazes

- Proving things is an NP problem, making solving this question more complex.

## TODO Exercises

Do the following Chapter 16 Exercises:

### 16.7.2.1

A SAT instance is given below in the DIMACS format.
1. What is the CNF formula capture by this instance?
2. By inpection, answer whether the instance is satisfiable, and why
3. If it is not satisfiable, then what is the minimal number of rows that must be deleted before the instance becomes satisfiable? If these rows are not unique, list the first two possible such omissions (of sets of rows), starting from the top of the given listing.
4. Check your answer using CryptoMiniSat

1. $(x(1) \lor x(2) \lor x(3) \lor x(4) \lor x(5)) \land \\
 (x(1) \lor x(2) \lor x(3) \lor x(4) \lor \lnot x(5)) \land \\
 ... \land (\lnot x(1) \lor \lnot x(2) \lor \lnot x(3) \lor \lnot x(4) \lor \lnot x(5))$
2. It is not satisfiable, because the first and the last clauses are mutually exclusive. 
3. The minimum number of rows that can be removed to make it satisfiable is 1. By removing the first or last row it is satisfiable.
4. Checked, results are in the next cell.

s SATISFIABLE
c ------- FINAL TOTAL SEARCH STATS ---------
c restarts                 : 1           (0.00      confls per restart)
c blocked restarts         : 0           (0.00      per normal restart)
c decisions                : 0           (0.00      % random)
c propagations             : 0           
c decisions/conflicts      : 0.00        
c conflicts                : 0           
c conf lits non-minim      : 0           (0.00      lit/confl)
c conf lits final          : 0.00        
c cache hit re-learnt cl   : 0           (0.00      % of confl)
c red which0               : 0           (0.00      % of confl)
c props/decision           : 0.00        
c props/conflict           : 0.00        
c 0-depth assigns          : 2           (0.00      % vars)
c [occ-substr] long subBySub: 0 subByStr: 0 lits-rem-str: 14
c [scc] new: 1 BP 0M  T: 0.00
c Conflicts in UIP         : 0           
c Mem used                 : 0.00        MB
v -1 -2 -3 -4 -5 0

c A SAT instance in DIMACS format  
c Your task is to determine whether this instance is satisfiable  
c  
p cnf 5 32  
1 2 3 4 5 0  
1 2 3 4 -5 0  
1 2 3 -4 5 0  
1 2 3 -4 -5 0  
1 2 -3 4 5 0  
1 2 -3 4 -5 0  
1 2 -3 -4 5 0  
1 2 -3 -4 -5 0  
1 -2 3 4 5 0  
1 -2 3 4 -5 0  
1 -2 3 -4 5 0  
1 -2 3 -4 -5 0  
1 -2 -3 4 5 0  
1 -2 -3 4 -5 0  
1 -2 -3 -4 5 0  
1 -2 -3 -4 -5 0  
-1 2 3 4 5 0  
-1 2 3 4 -5 0  
-1 2 3 -4 5 0  
-1 2 3 -4 -5 0  
-1 2 -3 4 5 0  
-1 2 -3 4 -5 0  
-1 2 -3 -4 5 0  
-1 2 -3 -4 -5 0  
-1 -2 3 4 5 0  
-1 -2 3 4 -5 0  
-1 -2 3 -4 5 0  
-1 -2 3 -4 -5 0  
-1 -2 -3 4 5 0  
-1 -2 -3 4 -5 0  
-1 -2 -3 -4 5 0  
-1 -2 -3 -4 -5 0

### 16.7.2.2

Consider the set of undirected graphs $\langle G\rangle$ with a set of nodes $N$ and a set of edges $E\subseteq N $x$ N$ such that we can two-color the graph (meaning no two nodes connected by an edge have the same color).

All of the graphs in $\langle G\rangle$ can be reduced to bipartite graphs. We can achieve this by saying that all nodes colored with the first color are in the first set of the bipartite graph and the nodes with the other color are of the other set. Since no two nodes with the same color can share an edge this preserves the requirement of a bipartite graph.

### 16.7.2.8

A Hamiltonian cycle in a graph with respect to a given node *n* is a tour that begins at *n* and visits all other nodes exactly once, returning to *n*. In a 5-clique, how many distinct Hamiltonian cycles exist? How about in an *n*-clique?
$\frac{!(n-1)}{2}$

Because by definition an n-clique has every vertex connected to every other vertex in the clique, it is the same as a $k_n$ graph. 
Suppose we start at some vertex v, from that vertex we have $(n-1)$ choices of other verticies to go to. From that vertex we have $(n-2)$ choices and so on. Thus the total number of potential hamiltonian cycles is $(n-1)!$. However this will also include all cycles in their reverse direction. Thus the total number of distinct cycles is $\frac{!(n-1)}{2}$.

### 16.7.2.9

Define the language *HalfClique* to be the set of input encodings $\langle G \rangle$ such that $G$ is an undirected graph having a clique with at least $n/2$ nodes, where $n$ is the number of nodes in $G$. Show that *HalfClique* is NPC. *Hint: Mapping reduction from Clique.*

First let us prove that *HalfClique* can be verified in p time.

Suppose we have some potential clique which is represented as a set of verticies s. To verify that s is a clique of size n/2 we must first verify that it has n/2 nodes. This operation takes O(1) time. We must then verify that each node in s has an out degree of (n/2-1). This will take O((n/2-1)^2) time since s represents a complete graph it will have that many edges. Therefore verifying if s is a clique of size n/2 takes polynomial time.

Now let us prove that generating *HalfClique* takes more than p time.
We can easily reduce *HalfClique* to SAT. 
Let us represent G as a graph G=(V,E), where we have a subgraph $x_{iv}$ which is a clique of length n/2 where $x_{iv}$ represents the ith vertex in the clique which is v. $x_{iv}$ has the following constraints:
1. There is an ith vertex: for all $V_x \in V_v$ (some node is the ith node of the clique)
2.  For each i, j < n/2 & $i \neq j$ there is $\lnot x_{iv} \lor \lnot x_{jv}$ (meaning that no node vertex is both the ith and jth vertex of the clique)
3. For each r $\neq$ s  and i < j such that (i,j) is not an edge of G, $\lnot x_{ir} \lor \lnot x_{js}$ (If there is no edge from i to j then i and j cannot both be in the clique).

Because we can reduce *HalfClique* to SAT it is NP Hard, because it can also be verified in p time, *HalfClique* is NP-Complete.

### 16.7.2.10

The game of Sudoku has been shown to be in NPC. In practice, one can encode and solve Sudoku using SAT solvers. Study the Sudoku solver (MIT license) written by Nicholas Pilkington at https://gist.github.com/nickponline/9c91fe65fef5b58ae1b0 Test it on the instance provided as well as a few that you create. Note: This solver will need Python2 (or you may adapt it for Python3).

In [1]:
!pip install pycosat

     |████████████████████████████████| 71kB 3.3MB/s 
  Created wheel for pycosat: filename=pycosat-0.6.3-cp37-cp37m-linux_x86_64.whl size=143807 sha256=4e7b8db363e5e9effbc80fb51545a00ff16f8004c73d27c4c301dc44c0eee860
  Stored in directory: /root/.cache/pip/wheels/c4/67/ff/5570304e45814eccef48a3c69c3af25d0456ed3a34eddbbe38
Successfully built pycosat


In [ ]:
# Author: Nicholas Pilkington, 2015
# License: MIT
# Blog Post: https://nickp.svbtle.com/sudoku-satsolver
import pycosat

N = 9
M = 3

def exactly_one(variables):
    cnf = [ variables ]
    n = len(variables)

    for i in range(n):
        for j in range(i+1, n):
            v1 = variables[i]
            v2 = variables[j]
            cnf.append([-v1, -v2])

    return cnf

def transform(i, j, k):
    return i*N*N + j*N + k + 1

def inverse_transform(v):
    v, k = divmod(v-1, N)
    v, j = divmod(v, N)
    v, i = divmod(v, N)
    return i, j, k

if __name__ == '__main__':
    cnf = []

    # Cell, row and column constraints
    for i in range(N):
        for s in range(N):
            cnf = cnf + exactly_one([ transform(i, j, s) for j in range(N) ])
            cnf = cnf + exactly_one([ transform(j, i, s) for j in range(N) ])

        for j in range(N):
            cnf = cnf + exactly_one([ transform(i, j, k) for k in range(N) ])

    # Sub-matrix constraints
    for k in range(N):
        for x in range(M):
            for y in range(M):
                v = [ transform(y*M + i, x*M + j, k) for i in range(M) for j in range(M)]
                cnf = cnf + exactly_one(v)

    # See contribution from @GregoryMorse below
    cnf = { frozenset(x) for x in cnf }
    cnf = list(cnf)

    # A 16-constraint Sudoku
    constraints = [
        (0, 3, 7),
        (1, 0, 1),
        (2, 3, 4),
        (2, 4, 3),
        (2, 6, 2),
        (3, 8, 6),
        (4, 3, 5),
        (4, 5, 9),
        (5, 6, 4),
        (5, 7, 1),
        (5, 8, 8),
        (6, 4, 8),
        (6, 5, 1),
        (7, 2, 2),
        (7, 7, 5),
        (8, 1, 4),
        (8, 6, 3),

    ]

    constraints2 = [
        (0, 3, 9),
        (1, 0, 5),
        (2, 3, 4),
        (2, 4, 3),
        (2, 6, 7),
        (3, 8, 9),
        (4, 3, 2),
        (4, 5, 9),
        (5, 6, 1),
        (5, 7, 5),
        (5, 8, 8),
        (6, 4, 2),
        (6, 5, 1),
        (7, 2, 2),

    ]

    cnf = cnf + [[transform(z[0], z[1], z[2])-1] for z in constraints2]

    for solution in pycosat.itersolve(cnf):
        X = [ inverse_transform(v) for v in solution if v > 0]
        for i, cell in enumerate(sorted(X, key=lambda h: h[0] * N*N + h[1] * N)):
            print(cell[2]+1, end=" ")
            if (i+1) % N == 0: print("")

### 16.7.2.12

Show that a 3-CNF formula<br>
$\Phi= \phi_0 \land \phi_1 \land ... \phi_i \land \phi_{N-1}$<br>
is unsatisfiable *if and only if* **for any variable assignment**, there is one clause $\phi_j$ with all the literals are true and another clause $\phi_k$ with all the literals are false.<br>
**Illustration:** Consider<br>
$\Phi= (x+y+y)\cdot(x+!y+!y)\cdot(!x+y+y)\cdot(!x+!y+!y)$<br>
For any assignment (say $x=0,y=1$), we have one clause whose literals are all true, and one clause whose literals are all false (these are respectively $(!x+y+y)$ and $(x+!y+!y)$). If we leave out any one clause, that is not the case, as $\Phi$ becomes satisfiable. Now finish the proof.

Because $\Phi$ requires that every $\phi$ be true to be true then if $\phi_k$ is true than all literals must be false. If that is the case then $\phi_J$ must be false. Thus not all $\phi$ in $\Phi$ can be true, which makes $\Phi$ unsatisfiable.

## TODO Explore a Powerful Tool

Do the following NON-book exercise:

SAT, while being NP-Complete, is a "workhorse of a tool." This exercise asks you to get a taste of running the CryptoMinisat tool on a non-trivial SAT formula. Click on https://msoos.github.io/cryptominisat_web/ to access this tool. When it comes up, it has a prefilled formula. There is a Play button that you can click, whereupon it solves the SAT instance.

This exercise asks you to replace this SAT instance with something bigger: specifically, the Pigeonhole problem (hole6.cnf) from: https://people.sc.fsu.edu/~jburkardt/data/cnf/cnf.html

Just click the above link, get the hole6.cnf file, and plunk the CNF into the buffer.

Hit Play and report on the execution time (you can just look at your phone's clock). If under 2 seconds, say "negligible" for your answer!

How much time would such a problem take through brute-force enumeration of $2^n$ combinations on a computer that takes a microsecond per variable combination (the $n$ is the number of variables used in the Pigeonhole problem)? HINT: Here is how you read the contents of a CNF file:

```
c File:  hole6.cnf <--- these are comment lines - starts wth a "C"
c...
c
p cnf 42 133
-1     -7
-1     -13
...
0 0
<--- CRUCIAL !! Tells you there are 42 variables and 133 clauses
<--- This line says (!x1 + !x7). The "0" is just end-of-a-clause marker!
<--- This line says (!x1 + !x13)
 12     11     10     9      8      7    0 <--- This clause reads
                                          (x12 + x11 + x10 + x9 + x8 + x7)
```

Okay, now you have all the info you need to calculate the time it takes to enumerate $2^n$ combinations!!



Your (two-part) answer goes here.

1. CryptoMinisat runtime: negligable
2. $2^n$ runtime estimation: (2^42)/(10^6) $\approx$ 3.3 seconds

## TODO Read and Learn

List six facts that you found interesting about Boolean SAT in these articles:

https://cacm.acm.org/magazines/2009/8/34498-boolean-satisfiability-from-theoretical-hardness-to-practical-success/fulltext

and

https://en.wikipedia.org/wiki/Boolean_satisfiability_problem

Anything that interested you is fine – theoretical or practical. Please write 1-2 sentences per point that interested you.

- Boolean Satisfiability problem: Given a formula, find a satisfying assignment or prove that none exists. 
- SAT belongs to the class of problems knows as NP-complete. This means that the time to solve the problems rows exponentially with the size of the instance.
- SAT was the first problem that was proven to be NP-complete. There is no known algorithm that can be used to solve every SAT problem efficently. SAT-algorithms can solve problem instances.
- SAT is the abreviation of Boolean satisfiabiliy problem. This is the problem of determining if there exists an interpretation that satisfies a given Boolean formula.
- There are many special cases in SAT problems. Due to this, some SAT problems like to defind the notion of generalized conjunctive normal form formula.
- SAT problems are self-reducible - which means that each algorithm which correctly answers if an instance of SAT is solvable can be used to find a satisfying assignment.
